spec/models/group_user_spec.rb

```diff
@@ -224,4 +224,18 @@ def levels
       expect(group.group_users.find_by(user_id: user_2.id).first_unread_pm_at).to eq_time(10.minutes.ago)
     end
   end
+
+  describe '#destroy!' do
+    it 'removes `primary_group_id`, `flair_group_id` and `title` for user' do
```

Comment:
I'm not sure if this is correct because the `GroupUser` callbacks will only remove `primary_group_id` for the user. I think the relevant assertions should live in the test files of the model where the callbacks are present.

Reply:
I'm also confused now :) The bug we're trying to fix here is happening whenever the `GroupUser`'s `destroy` method is called. Even if I write this in the group model's spec then I have to check it against the `group_user.destroy!` method only. I guess it's correct since I want to make sure  `group_user.destroy!` is removing the values from `user.flair_group_id` and `user.title`.

Reply:
Sorry where are the callbacks being set for `User#flair_group_id` and `User#title` when `group_user.destroy!` is called?

Reply:
[`user.update_attribute(:primary_group_id, nil)`](https://github.com/discourse/discourse/blob/09593edf03d036dd602a16db9d750d18f49a09d5/app/models/group_user.rb#L88) method will trigger the `before_save` callback in `user` model which will run the [`match_primary_group_changes`](https://github.com/discourse/discourse/blob/09593edf03d036dd602a16db9d750d18f49a09d5/app/models/user.rb#L137) method. It removes these field values if possible.

Reply:
> callback in `user` model which will run the [`match_primary_group_changes`](https://github.com/discourse/discourse/blob/09593edf03d036dd602a16db9d750d18f49a09d5/app/models/user.rb#L137) method. It removes these field values if possible.

In this case, I think those assertion should be on the `User` model and not be tested as a side effect of the `GroupUser` model.

Reply:
I understand what you're saying. Still, I want to make sure the `before_save` method is being triggered when run the `destroy` method for group_user. Let me check.

Reply:
Do you mean the `after_destroy :remove_primary_group` callback ?

Reply:
No, I mean the `before_save` callback of the user model. https://github.com/discourse/discourse/pull/14367/commits/b81d8ba9bd676f44a3b7afeb4f7f8028dce22868 I hope this is a correct approach here.

Reply:
I'll split it up to two tests here. A callback on the `User` model is not a concern for the `GroupUser` model to test for because the `GroupUser`'s model only concern is to run a callback that updates `User#primary_group_id`. What happens due to a change in `User#primary_group_id` is that of the `User`'s model concern because it has a callback that is observing for that change.

Reply:
Let's assume I create two tests here. One test is checking the removal of `primary_group_id`. And another one in the user model is checking the removal of `flair_group_id` when the `primary_group_id` is changed.

 Now if someone used `DB.exec` method again [here](https://github.com/discourse/discourse/pull/14367/files#diff-c78cab2f58ea573db792df9ad80c8f1f7bbcd71e58ffab470f4888d232d48d8eL86-R88) or used `user.update_column(:primary_group_id, nil)` method, then both the tests will pass but the bug I'm trying to fix in this PR will popup again.

Reply:
ahh ic what you mean now. This is certainty tricky and if I have a choice I would avoid all these nested callbacks we have on our models now. Anyway, let's just go with the simple change here first and we can revisit in the future.

Reply:
Yes. I even want to disable all other ways to remove/join a user in a group except `group.remove(user)` and `group.add(user)`. Things getting complex since we have multiple ways to handle user's membership in a group.

Reply:
Yea that was my thought too. There should only be one way to add or remove a user from a group just like how we do it for `PostCreator`.